File: 53. Maximum Subarray/maxSubArray.js
/**
 * @param {number[]} nums
 * @return {number}
 */
const maxSubArray = (nums) => {
  let maxSubArraySum = nums[0];
  let currentSum = nums[0];

  for (let i = 1; i < nums.length; i++) {
    currentSum = Math.max(nums[i], currentSum + nums[i]);
    maxSubArraySum = Math.max(maxSubArraySum, currentSum);
  }

  return maxSubArraySum;
};


const nums = [-2, 1, -3, 4, -1, 2, 1, -5, 4];
console.log('maxSubArray: ', maxSubArray(nums));

File: 53. Maximum Subarray/maxSubArray.py
from typing import List

class Solution:
  def maxSubArray(self, nums: List[int]) -> int:
    maxSubArraySum = nums[0]
    currentSum = 0

    for n in nums:
      currentSum = max(n, currentSum + n)
      maxSubArraySum = max(maxSubArraySum, currentSum)

    return maxSubArraySum


nums = [-2, 1, -3, 4, -1, 2, 1, -5, 4]
solution = Solution()
print('maxSubArray: ', solution.maxSubArray(nums))
